### PR TITLE
Adds Failure Component to Scaffolded Cells

### PR DESCRIPTION
**Pull request by dthyresson**
Fixes https://github.com/redwoodjs/redwood/issues/2853

The components have beed added to the scaffold templates, snapshots updated and e2e passes.

However, want to add an e2e step that handles the error:

* in a query
* in a mutation

The PR

* adds option to set secure service envar when standing up new RW app for testing
* defaults to enable secure services
* adds scenario to test for a query failure due to not being authenticated
* in the tutorial, adds codemods to alter `auth.js` to raise a ForbibbenError
* updates posts service to require auth
* tests for cell failure error message

Note: did not do mutation as redundant but could set before resolver case for just get all and update and test those.

Let me know if think is needed.

### CODE REVIEW

**Comment by dac09** [comment]
@dthyresson is there anyway you could modify the scaffold command to use the cell generator underneath? It'll save us a lot of duplication - and the cell generator handles lists, types, etc. much better

**Comment by dthyresson** [comment]
> @dthyresson is there anyway you could modify the scaffold command to use the cell generator underneath? It'll save us a lot of duplication - and the cell generator handles lists, types, etc. much better

Maybe, but recall a cell generatre generate one component: a single cell or a cell for list of items. The Scaffold generates a view one, a view list, and edit.

And the cell content has the markup and layout and css for the forms to show, edit, delete etc.
image

image


So -- it would be tough I think to do that.

**Comment by dac09** [comment]
I understand there's more files in scaffold, sorry wasn't clearer. What I meant was:

- For the Cell, re-use the cell generator. 
- Likewise for the service, re-use the service generator if possible. 
- For any other files leave them as is because its scaffold specific

tldr; don't duplicate generators

**Comment by Tobbe** [comment]
@dac09 I agree it would be awesome to reuse the cell generator. But how
would you get the required css classes etc in there?

**Comment by dac09** [comment]
> would you get the required css classes etc in there?
It seems to me that the only Cell template in Scaffolds that's different is the EditNameCell. I think that's probably totally fine to leave it in here. But the other ones 

`NameCell.js.template` -> use cell generator with `list` false 
`NamesCell.js.template` -> use cell generator with `list` true

**Comment by dthyresson** [comment]
> `NameCell.js.template` -> use cell generator with `list` false
> `NamesCell.js.template` -> use cell generator with `list` true

@dac09 the scaffold `NamesCell` needs the following

```js
export const Empty = () => {
  return (
    <div className="rw-text-center">
      {'No ${pluralCamelName} yet. '}
      <Link
        to={routes.${newRouteName}()}
        className="rw-link"
      >
        {'Create one?'}
      </Link>
    </div>
  )
}
```

It's Empty state creates a button to Create a new widget.

And it's Success:

```js
export const Success = ({ ${pluralCamelName} }) => {
  return <${pluralPascalName} ${pluralCamelName}={${pluralCamelName}} />
}
```

uses a component to render a table with a row and Show, Edit, Delete buttons.

So, the content of the template is not the same.

**Comment by dac09** [comment]
> So, the content of the template is not the same.

Sad :'(. 

Cool - so for the magical typescript stuff you'll probably need to add these lines:

1. At the top
```
import type { Find${singularPascalName}ById } from 'types/graphql'
import type { CellSuccessProps, CellFailureProps } from '@redwoodjs/web'
```

2. For Failure 
```
 const Failure = ({ error }: CellFailureProps)
```

3. For Success
```
export const Success = ({ ${camelName} }: CellSuccessProps<Find${singularPascalName}ById>) => {
```

**Comment by dthyresson** [comment]
> so for the magical typescript stuff

Would we also rename these to `.ts` or `tsx`?

The scaffold has other components, too, to Typsescript-ify (like the Form, etc).

**Comment by dac09** [comment]
Yeah rename to `.tsx`- I think for Forms, they don't need explicit types really - all the types there can be inferred.

**Comment by dthyresson** [comment]
Note making a separate issue for making scaffolds Typescript.

**Review by dac09 — APPROVED** [review]
LGTM!
